src/ofdm.c

```diff
@@ -1249,45 +1275,72 @@ static void ofdm_demod_core(struct OFDM *ofdm, int *rx_bits) {
+         * Use all pilots normally, results in low SNR performance,
+         * but will fall over in high Doppler propagation
+         *
+         * Basically we divide the Nc+2 pilots into groups of 3
+         * Then average the phase surrounding each of the data symbols.
+         */
+        for (i = 1; i < (ofdm_nc + 1); i++) {
+            complex float symbol[3];
 
-    for (i = 1; i < (ofdm_nc + 1); i++) {
-        complex float symbol[3];
+            for (j = (i - 1), k = 0; j < (i + 2); j++, k++) {
+                symbol[k] = ofdm->rx_sym[1][j] * conjf(ofdm->pilots[j]); /* this pilot conjugate */
+            }
 
-        for (j = (i - 1), k = 0; j < (i + 2); j++, k++) {
-            symbol[k] = ofdm->rx_sym[1][j] * conjf(ofdm->pilots[j]); /* this pilot conjugate */
-        }
-        
-        aphase_est_pilot_rect = vector_sum(symbol, 3);
+            aphase_est_pilot_rect = vector_sum(symbol, 3);
 
-        for (j = (i - 1), k = 0; j < (i + 2); j++, k++) {
-            symbol[k] = ofdm->rx_sym[ofdm_ns + 1][j] * conjf(ofdm->pilots[j]); /* next pilot conjugate */
-        }
+            for (j = (i - 1), k = 0; j < (i + 2); j++, k++) {
+                symbol[k] = ofdm->rx_sym[ofdm_ns + 1][j] * conjf(ofdm->pilots[j]); /* next pilot conjugate */
+            }
 
-        aphase_est_pilot_rect = aphase_est_pilot_rect + vector_sum(symbol, 3);
+            aphase_est_pilot_rect = aphase_est_pilot_rect + vector_sum(symbol, 3);
 
-        /* use next step of pilots in past and future */
+            /* use pilots in past and future */
 
-        for (j = (i - 1), k = 0; j < (i + 2); j++, k++) {
-            symbol[k] = ofdm->rx_sym[0][j] * conjf(ofdm->pilots[j]); /* previous pilot */
+            for (j = (i - 1), k = 0; j < (i + 2); j++, k++) {
+                symbol[k] = ofdm->rx_sym[0][j] * conjf(ofdm->pilots[j]); /* previous pilot */
+            }
+
+            aphase_est_pilot_rect = aphase_est_pilot_rect + vector_sum(symbol, 3);
+
+            for (j = (i - 1), k = 0; j < (i + 2); j++, k++) {
+                symbol[k] = ofdm->rx_sym[ofdm_ns + 2][j] * conjf(ofdm->pilots[j]); /* future pilot */
+            }
+
+            aphase_est_pilot_rect = aphase_est_pilot_rect + vector_sum(symbol, 3);
+            aphase_est_pilot[i] = cargf(aphase_est_pilot_rect);
+
+            /* amplitude is estimated over 12 pilots */
+
+            aamp_est_pilot[i] = cabsf(aphase_est_pilot_rect / 12.0f);
         }
+    } else {
+        /*
+         * Use only symbols at 'this' and 'next' to quickly track changes
+         * in phase due to high Doppler spread in propagation.
+         * 
+         * As less pilots are averaged, low SNR performance will be poorer
+         */
+        complex float symbol[ofdm_nc];
 
-        aphase_est_pilot_rect = aphase_est_pilot_rect + vector_sum(symbol, 3);
+        for (j = 1, k = 0; j <= ofdm_nc; j++, k++) {
+            symbol[k] = ofdm->rx_sym[1][j] * conjf(ofdm->pilots[j]); /* this pilot conjugate */
```

Comment:
@srsampson I think the for() loop indexes should be these same as above?

```for (j = (i - 1), k = 0; j < (i + 2); j++, k++) {```

Also, missing the outer loop ```for(i ..```

Reply:
aamp_est_pilot[i] in the high_doppler block was an error. Not sure how to fix it though, as code further down expects an array and not a scalar.

Reply:
The Octave code is confusing. It uses the variable 'c' which is defined as 2:18

Reply:
Ahh sorry I think my previous comment was wrong, as we are not averaging 3 pilots. Yes it's a bit tricky for sure:

```
achannel_est_rect(c) =  sum(rx_sym(2,c)*pilots(c)');      % frame    
achannel_est_rect(c) += sum(rx_sym(2+Ns,c)*pilots(c)');   % frame+1
```
This is inside the for c loop in octave, so iterating over each carrier.  I note this is a scalar multiply, so it's not working on vectors of 3.  In high doppler mode, we're down to just two pilots at the start and end of a frame for a single carrier.

Reply:
Yes, I see where I messed a bunch of that up. Will scrape that off and try again.

Reply:
I'd suggest a similar ```for(i``` loop to the code above, compute one ```aphase_est_pilot_rect``` per carrier, just using less points. I think ATM:
```aphase_est_pilot_rect = vector_sum(symbol, ofdm_nc)```
is averaging over all carriers.  I need to head off the $dayjob now, I'm sure you'll get there.  It's very close :+1:

Reply:
The pull was probably premature, but at least we had a nice chat. I'll do another pull after testing the code. Have fun!

Reply:
The latest code was updated and the data is:

````
$ nc=17; ./ofdm_mod --in /dev/zero --testframes 300 --nc $nc | ./cohpsk_ch - - -100 --Fs 8000 -f 10 --ssbfilt 1 --fast | ./ofdm_demod --high_doppler 0 --out /dev/null --testframes --nc $nc --verbose 1
cohpsk_ch ----------------------------------------------------------------------------------
interleave_frames: 1
high doppler mode: disabled
Fs: 8000 NodB: -100.00 foff: 10.00 Hz fading: 1 inclip: 1.00 ssbfilt: 1
cohpsk_ch ----------------------------------------------------------------------------------
   ct_est:  420 foff_est:  7.7 timing_valid: 1 timing_mx: 0.4828
cohpsk_ch -----------------------------------------------------------------------------
SNR3k(dB): 78.25 C/No: 113.0 PAPR:  9.7 
cohpsk_ch -----------------------------------------------------------------------------
BER......: 0.0304 Tbits: 88774 Terrs:  2701
BER2.....: 0.0279 Tbits: 84014 Terrs:  2341

$ nc=17; ./ofdm_mod --in /dev/zero --testframes 300 --nc $nc | ./cohpsk_ch - - -100 --Fs 8000 -f 10 --ssbfilt 1 --fast | ./ofdm_demod --high_doppler 1 --out /dev/null --testframes --nc $nc --verbose 1
cohpsk_ch ----------------------------------------------------------------------------------
Fs: 8000 NodB: -100.00 foff: 10.00 Hz fading: 1 inclip: 1.00 ssbfilt: 1
cohpsk_ch ----------------------------------------------------------------------------------
interleave_frames: 1
high doppler mode: enabled
   ct_est:  420 foff_est:  7.7 timing_valid: 1 timing_mx: 0.4828
cohpsk_ch -----------------------------------------------------------------------------
SNR3k(dB): 78.25 C/No: 113.0 PAPR:  9.7 
cohpsk_ch -----------------------------------------------------------------------------
BER......: 0.0234 Tbits: 88774 Terrs:  2077
BER2.....: 0.0214 Tbits: 84014 Terrs:  1801
```

**Which still seems too high.**

Reply:
I apologize the Octave code is a little strange for me... It looks like it is using sum() but the intent maybe is to not average the selected carrier with the the surrounding left and right carrier. If that is the case, we don't need sum() at all, and can just add the two single carriers from this and next rows, and then divide by 2.0.

If that is the case, it results in:

```
$ nc=17; ./ofdm_mod --in /dev/zero --testframes 300 --nc $nc | ./cohpsk_ch - - -100 --Fs 8000 -f 10 --ssbfilt 1 --fast | ./ofdm_demod --high_doppler 1 --out /dev/null --testframes --nc $nc --verbose 1
cohpsk_ch ----------------------------------------------------------------------------------
Fs: 8000 NodB: -100.00 foff: 10.00 Hz fading: 1 inclip: 1.00 ssbfilt: 1
cohpsk_ch ----------------------------------------------------------------------------------
interleave_frames: 1
high doppler mode: enabled
   ct_est:  420 foff_est:  7.7 timing_valid: 1 timing_mx: 0.4828
cohpsk_ch -----------------------------------------------------------------------------
SNR3k(dB): 78.25 C/No: 113.0 PAPR:  9.7 
cohpsk_ch -----------------------------------------------------------------------------
BER......: 0.0165 Tbits: 88774 Terrs:  1466
BER2.....: 0.0153 Tbits: 84014 Terrs:  1282
$ nc=17; ./ofdm_mod --in /dev/zero --testframes 300 --nc $nc | ./cohpsk_ch - - -100 --Fs 8000 -f 10 --ssbfilt 1 --slow | ./ofdm_demod --high_doppler 1 --out /dev/null --testframes --nc $nc --verbose 1
cohpsk_ch ----------------------------------------------------------------------------------
Fs: 8000 NodB: -100.00 foff: 10.00 Hz fading: 2 inclip: 1.00 ssbfilt: 1
cohpsk_ch ----------------------------------------------------------------------------------
interleave_frames: 1
high doppler mode: enabled
   ct_est:  421 foff_est: 11.8 timing_valid: 1 timing_mx: 0.4990
cohpsk_ch -----------------------------------------------------------------------------
SNR3k(dB): 78.25 C/No: 113.0 PAPR:  9.7 
cohpsk_ch -----------------------------------------------------------------------------
BER......: 0.0006 Tbits: 88774 Terrs:    57
BER2.....: 0.0006 Tbits: 84014 Terrs:    50
```
and the code is:

```
        } else {
            /*
             * Use only symbols at 'this' and 'next' to quickly track changes
             * in phase due to high Doppler spread in propagation.
             * 
             * As less pilots are averaged, low SNR performance will be poorer
             */
            aphase_est_pilot_rect = ofdm->rx_sym[1][i] * conjf(ofdm->pilots[i]); /* this pilot conjugate */
            aphase_est_pilot_rect += ofdm->rx_sym[ofdm_ns + 1][i] * conjf(ofdm->pilots[i]); /* next pilot conjugate */

            aphase_est_pilot[i] = cargf(aphase_est_pilot_rect);

            /* amplitude is estimated over 2 pilots */

            aamp_est_pilot[i] = cabsf(aphase_est_pilot_rect / 2.0f);
        }
```

Reply:
Good Job Steve :+1:  No need to apologise, it's a tricky bit of code.  I probably had too much coffee when I wrote the Octave version : -)

I noticed that cohpsk_ch was stopping the simulation prematurely (#bits tested wasn't changing with --testframes) , so have built some bigger fade files:
 ```
octave:3> cohpsk_ch_fading("../raw/fast_fading_samples.float", 8000, 1.0, 8000*500)
octave:4> cohpsk_ch_fading("../raw/slow_fading_samples.float", 8000, 0.1, 8000*500)
```
This was a trap with cohpsk_ch, it ends silently when the fade file is finished.  I've added a warning.

On 300 seconds (500,000 bits) --fast I'm now getting BER of 0.016/0.01 with high_doppler off/on , not as great as I'd hoped.  When I use --log and ofdm_demod_c to plot the scatter diagram doesn't look too good, so that's a hint. Maybe I'll head back to the Octave simulation and repeat the tests.

There's a spec somewhere for the length of simulation you need for reliable results at a given Doppler bandwidth.

Reply:
@srsampson - so just to log our current results here for future work:
1. I have noticed a similar error floor for 1Hz Doppler with the ofdm_dev.m Octave simulation.
1. The noise on the scatter diagram (thickness of cross for fading channels) gets wider with increasing Doppler, so 0.1Hz is fine, and progressively worse for 0.5., 1, 2 Hz etc
1. In Octave, increasing the pilot insertion rate Ns helps.
1. At Nc=31 --high_doppler=0/1 I get 0.026/0.02.  Curiously, with with ssb_filt=0 in cohpsk_ch, I get a very high BER (0.16) with --high_doppler=0.

It would be useful to get to the bottom of this error floor/Doppler issue one day.  We may need some more R&D with the Octave sims, looking into why the phase estimation leads to noise on the scatter plot.  2% error for the Nc=31 waveform is a worry, as it's eating up a lot of the LDPC codes error correcting capacity.  Might be a good one to work on with Mike Riley.

Reply:
@srsampson when you're happy we can merge with master.

Reply:
I did a compile and ran "make test" and all passes. Also ran the fast and slow script to make sure the numbers were the same. It's ready for a merge.

Reply:
Great - thanks Steve - I think @db4ple has configured Travis to run the tests automatically as well, "show all checks" link below.